clang/test/SemaSYCL/stall_enable_device.cpp

```diff
@@ -0,0 +1,61 @@
+// RUN: %clang_cc1 %s -fsyntax-only -internal-isystem %S/Inputs -fsycl-is-device -Wno-sycl-2017-compat -DTRIGGER_ERROR -verify
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -fsyntax-only -ast-dump -Wno-sycl-2017-compat %s | FileCheck %s
+
+// Test that checks [[intel::use_stall_enable_clusters]] attribute support on function.
+// The metadata to be attached to the functionDecl that the attribute is applied to.
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
+
+// Test attribute is presented on function metadata.
+[[intel::use_stall_enable_clusters]] void test() {}
+// CHECK: FunctionDecl{{.*}}test
+// CHECK: SYCLIntelUseStallEnableClustersAttr
+
+// Tests for incorrect argument values for Intel FPGA use_stall_enable_clusters function attribute.
+#ifdef TRIGGER_ERROR
+[[intel::use_stall_enable_clusters(1)]] void test1() {} // expected-error{{'use_stall_enable_clusters' attribute takes no arguments}}
+[[intel::use_stall_enable_clusters]] int test2;         // expected-error{{'use_stall_enable_clusters' attribute only applies to functions}}
+#endif
+
+// Test attribute is presented on function call operator (of a function object).
+struct FuncObj {
+  [[intel::use_stall_enable_clusters]] void operator()() const {}
+  // CHECK: CXXRecordDecl{{.*}}implicit struct FuncObj
+  // CHECK-NEXT: CXXMethodDecl{{.*}}used operator() 'void () const'
+  // CHECK-NEXT-NEXT:SYCLIntelUseStallEnableClustersAttr
+};
+
+// Test attribute is presented on lambda function.
+void test3() {
+  auto lambda = []() [[intel::use_stall_enable_clusters]]{};
+  lambda();
+  // CHECK: FunctionDecl{{.*}}test3
+  // CHECK: LambdaExpr
+  // CHECK: SYCLIntelUseStallEnableClustersAttr
+}
+
+int main() {
+  q.submit([&](handler &h) {
+    // Test attribute is not propagated to the kernel.
+    // CHECK-LABEL: FunctionDecl {{.*}}test_kernel1
+    // CHECK-NOT:   SYCLIntelUseStallEnableClustersAttr {{.*}}
+    h.single_task<class test_kernel1>(
+        FuncObj());
+
+    // Test attribute does not present on kernel.
+    // CHECK-LABEL: FunctionDecl {{.*}}test_kernel2
+    // CHECK-NOT:   SYCLIntelUseStallEnableClustersAttr {{.*}}
+    h.single_task<class test_kernel2>(
+        []() [[intel::use_stall_enable_clusters]]{});
```

Comment:
The attribute will be present on LambdaExpr here right?

Reply:
Yes, the attribute will be present on LambdaExpr.

Here is the Sema test i added
```


void test3() {  
auto lambda = []() [[intel::use_stall_enable_clusters]]{};  
lambda();  
// CHECK: FunctionDecl{{.*}}test3  
// CHECK: LambdaExpr  
// CHECK: SYCLIntelUseStallEnableClustersAttr

```
The attribute does not get propagated to the kernel, so the attribute will not be present on LambdaExpr called through kernel.

Reply:
@elizabethandrews , does this explain to your question?

Reply:
Sorry I am a bit confused. Can you paste the AST for this code in Github comments here?

Reply:
````
|-FunctionDecl 0x55af0f6ad790 </test/SemaSYCL/stall_enable_device.cpp:51:9, <invalid sloc>> col:9 _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE_clES2_E12test_kernel2 'void ()'
| |-CompoundStmt 0x55af0f6adad0 <<invalid sloc>>
| | |-DeclStmt 0x55af0f6ad9d0 </test/SemaSYCL/Inputs/sycl.hpp:231:23>
| | | `-VarDecl 0x55af0f6ad928 </test/SemaSYCL/stall_enable_device.cpp:51:9, 
/test/SemaSYCL/Inputs/sycl.hpp:231:23> 
/test/SemaSYCL/stall_enable_device.cpp:51:9 used '(lambda at 
/test/SemaSYCL/stall_enable_device.cpp:51:9)' cinit
| | |   `-InitListExpr 0x55af0f6ad990 </test/SemaSYCL/Inputs/sycl.hpp:231:23> '(lambda at /est/SemaSYCL/stall_enable_device.cpp:51:9)'
| | `-CompoundStmt 0x55af0f6adab8 <col:72, line:233:1>
| |   `-CXXOperatorCallExpr 0x55af0f6ada88 <line:232:3, col:14> 'void':'void' '()'
| |     |-ImplicitCastExpr 0x55af0f6ada58 <col:13, col:14> 'void (*)() const' <FunctionToPointerDecay>
| |     | `-DeclRefExpr 0x55af0f6ada38 <col:13, col:14> 'void () const' lvalue CXXMethod 0x55af0f6a5ee0 'operator()' 'void () const'
| |     `-ImplicitCastExpr 0x55af0f6ada70 <col:3> 'const (lambda at 
/test/SemaSYCL/stall_enable_device.cpp:51:9)' lvalue <NoOp>
| |       `-DeclRefExpr 0x55af0f6ada18 <col:3> '(lambda at /test/SemaSYCL/stall_enable_device.cpp:51:9)' lvalue Var 0x55af0f6ad928 '' '(lambda at /test/SemaSYCL/stall_enable_device.cpp:51:9)'
| |-OpenCLKernelAttr 0x55af0f6ad830 <<invalid sloc>> Implicit
| |-AsmLabelAttr 0x55af0f6ad888 <<invalid sloc>> Implicit "_ZTSZZ4mainENKUlRN2cl4sycl7handlerEE_clES2_E12test_kernel2"
| |-ArtificialAttr 0x55af0f6ad900 <<invalid sloc>> Implicit
| `-SYCLKernelAttr 0x55af0f6adaf0 <<invalid sloc>> Implicit
```

Reply:
I had an offline discussion with Soumi, who confirmed this attribute does not need to be attached to kernel and is basically ignored when used in this syntax.